gnc/navigator_perception/scan_the_code_detector.cc

```diff
@@ -0,0 +1,281 @@
+                                          ){
+
+    // Prevent segfault if service is called before we get valid img_msg_ptr's
+    if (model_fitter->left_most_recent.image_msg_ptr == NULL || model_fitter->right_most_recent.image_msg_ptr == NULL) {
+      ROS_WARN("Torpedo Board Detector: Image Pointers are NULL.");
+      throw "ROS ERROR";
+    }
+    double sync_thresh = 0.5;
+
+    // Get the most recent frames and camera info for both cameras
+    cv_bridge::CvImagePtr input_bridge;
+    try {
+
+      left_mtx.lock();
+      right_mtx.lock();
+
+      Matx34d left_cam_mat = left_cam_model.fullProjectionMatrix();
+
+      cout << "A = "<< endl << " "  << left_cam_mat << endl << endl;
+
+      // Left Camera
+      input_bridge = cv_bridge::toCvCopy(model_fitter->left_most_recent.image_msg_ptr,
+                                         sensor_msgs::image_encodings::BGR8);
+      current_image_left = input_bridge->image;
+      left_cam_model.fromCameraInfo(model_fitter->left_most_recent.info_msg_ptr);
+      resize(current_image_left, processing_size_image_left, Size(0, 0),
+                 image_proc_scale, image_proc_scale);
+      if (current_image_left.channels() != 3) {
+        ROS_ERROR("The left image topic does not contain a color image.");
+        throw "ROS ERROR";
+      }
+
+      // Right Camera
+      input_bridge = cv_bridge::toCvCopy(model_fitter->right_most_recent.image_msg_ptr,
+                                         sensor_msgs::image_encodings::BGR8);
+      current_image_right = input_bridge->image;
+      right_cam_model.fromCameraInfo(model_fitter->right_most_recent.info_msg_ptr);
+      resize(current_image_right, processing_size_image_right, Size(0, 0),
+                 image_proc_scale, image_proc_scale);
+      if (current_image_right.channels() != 3) {
+        ROS_ERROR("The right image topic does not contain a color image.");
+        throw "ROS ERROR";
+      }
+
+      left_mtx.unlock();
+      right_mtx.unlock();
+
+    } catch (const exception &ex) {
+      ROS_ERROR("[scan_the_code] cv_bridge: Failed to convert images");
+      left_mtx.unlock();
+      right_mtx.unlock();
+      throw "ROS ERROR";
+    }
+
+    // Enforce approximate image synchronization
+    double left_stamp, right_stamp;
+    left_stamp = model_fitter->left_most_recent.image_msg_ptr->header.stamp.toSec();
+    right_stamp = model_fitter->right_most_recent.image_msg_ptr->header.stamp.toSec();
+    double sync_error = fabs(left_stamp - right_stamp);
+    stringstream sync_msg;
+    sync_msg << "Left and right images were not sufficiently synchronized"
+             << "\nsync error: " << sync_error << "s";
+    if (sync_error > sync_thresh) {
+      ROS_WARN(sync_msg.str().c_str());
+      throw "Sync Error";
+    }
+}
+
+void ScanTheCodeDetector::init_ros(stringstream& log_msg){
+    using ros::param::param;
+    int tab_sz = 4;
+    // Default parameters
+    string img_topic_left_default = "/stereo/left/image_rect_color/";
```

Comment:
Most of the defaults are different with the navigator cameras, just fyi if you're not launching and setting these param